### PR TITLE
Integrate Micrometer with WebSockets Next

### LINKED ISSUE #39143: WebSockets Next: OTel integration
### Description

Follow-up of https://github.com/quarkusio/quarkus/pull/39142.

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by michalvavrik**
- closes: https://github.com/quarkusio/quarkus/issues/39143

### CODE REVIEW

**Comment by michalvavrik** [comment]
Interesting that `MicrometerWebSocketsOnBinaryMessageTest.testClientEndpoint_MultiTextReceived_MultiTextSent` is flaky. IIRC I cut down timeout to match the timeout used by WS Next OTel tests. Either I need to increase it or tune some setting. I'll have a look when I will be addressing reviewer requests for change.

File: extensions/websockets-next/deployment/src/main/java/io/quarkus/websockets/next/deployment/WebSocketProcessor.java [hunk 1]
```diff
@@ -685,6 +699,10 @@ private static boolean isTracesSupportEnabled(Capabilities capabilities) {
         return capabilities.isPresent(Capability.OPENTELEMETRY_TRACER);
     }
 
+    private static Boolean isMetricsSupportEnabled(Optional<MetricsCapabilityBuildItem> metricsCapability) {
```
Comment (mkouba) [inline extensions/websockets-next/deployment/src/main/java/io/quarkus/websockets/next/deployment/WebSocketProcessor.java:702]:
```suggestion
    private static boolean isMetricsSupportEnabled(Optional<MetricsCapabilityBuildItem> metricsCapability) {
```

File: extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/SendingInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import io.vertx.core.buffer.Buffer;
+
+/**
+ * Intercepts text or binary send from the {@link io.quarkus.websockets.next.runtime.WebSocketConnectionBase} connection.
+ */
+public sealed interface SendingInterceptor permits MetricsSendingInterceptor {
+
+    // following methods should mirror WebSocketConnectionBase sending methods
+
+    Runnable onSend(String text);
```
Comment (mkouba) [inline extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/SendingInterceptor.java:12]: I think that it would make a little more sense to change these methods to return `void` and let the call site create the `Runnable` where needed (ie. when the reactive pipeline is created).

**Review by mkouba — COMMENTED** [review]
Looks very good and the test coverage is really impressive...

File: extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/SendingInterceptor.java (hunk 2 above)
Comment (michalvavrik) [reply extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/SendingInterceptor.java]: Right, I think you are right, fixed.

**Comment by michalvavrik** [comment]
Ah, I forgot about that flakiness (it never fails locally), I'll address it now.

File: extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/TelemetryConstants.java [hunk 3]
```diff
@@ -13,4 +13,85 @@ private TelemetryConstants() {
...
+
+    /**
+     * Counts all the WebSockets client closed connections.
+     */
+    public static final String SERVER_CONNECTION_CLOSED = "quarkus.websockets.server.connections.closed";
+
+    /**
+     * Counts all the WebSockets server endpoint errors.
+     */
+    public static final String SERVER_MESSAGES_COUNT_ERRORS = "quarkus.websockets.server.messages.count.errors";
```
Comment (mkouba) [inline extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/TelemetryConstants.java:49]: I wonder if this should be something like `SERVER_ENDPOINT_COUNT_ERRORS = "quarkus.websockets.server.endpoint.count.errors" ` instead because it's not only about messages but about endpoint logic in general (IIUIC it mirros the `@OnError` functionality).

File: extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/TelemetryConstants.java [hunk 4]
```diff
@@ -13,4 +13,85 @@ private TelemetryConstants() {
...
+
+    /**
+     * Counts all the WebSockets client endpoint errors.
+     */
+    public static final String CLIENT_MESSAGES_COUNT_ERRORS = "quarkus.websockets.client.messages.count.errors";
```
Comment (mkouba) [inline extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/TelemetryConstants.java:54]: `CLIENT_ENDPOINT_COUNT_ERRORS = "quarkus.websockets.client.endpoint.count.errors"`??

File: extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/TelemetryConstants.java [hunk 5]
```diff
@@ -13,4 +13,85 @@ private TelemetryConstants() {
...
+
+    /**
+     * Number of messages sent and received by server endpoints.
+     */
+    public static final String SERVER_COUNT = "quarkus_websockets_server_count";
```
Comment (mkouba) [inline extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/TelemetryConstants.java:59]: Why do we sometimes use dots and sometimes underscores as delimiters in metric names?

File: extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/TelemetryConfig.java [hunk 6]
```diff
@@ -16,4 +16,12 @@ public interface TelemetryConfig {
     @WithDefault("true")
     boolean tracesEnabled();
 
+    /**
+     * If collection of WebSocket metrics is enabled.
+     * Only applicable when the Micrometer extension is present.
+     */
+    @WithName("metrics.enabled")
+    @WithDefault("true")
+    boolean metricsEnabled();
```
[the remark this answers is not in the document]
Comment (mkouba) [reply extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/TelemetryConfig.java:25]: That is a good point. Is it ok that traces are enabled by default?

[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/TelemetryConfig.java:25]: yes, that's the usual because it's more useful and has less memory impact than metrics.

[the remark this answers is not in the document]
Comment (michalvavrik) [reply extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/TelemetryConfig.java:25]: I'll change it and add a test that makes asserts no metrics are collected when disabled. I forgot to add it.

File: extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/TelemetryConstants.java (hunk 3 above)
Comment (michalvavrik) [reply extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/TelemetryConstants.java:49]: Yes, I agree with that. I think I called it differently before but changed it based on suggestions. They are not only error messages. I'll change it. We will see what @brunobat thinks.

Comment (michalvavrik) [reply extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/TelemetryConstants.java]: Done.

File: extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/TelemetryConstants.java (hunk 4 above)
Comment (michalvavrik) [reply extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/TelemetryConstants.java]: Done.

**Comment by mkouba** [comment]
> It's bit difficult as `MicrometerWebSocketsOnTextMessageTest` haven't failed locally yet, but I'll try to read code in next 2 days and figure why it could be flaky. There is probably wrong expected count somewhere.

I can try it locally too...

**Comment by michalvavrik** [comment]
> I can try it locally too...

Thanks, I'll call on you if I fail. It is impossible to  reproduce, but all that needs to be done is to check line by line and check expected numbers with what is actually going on. Also there is the fact that it doesn't happen for the binary. I should have time this evening. I'll let you know tomorrow if anything is necessary.

**Comment by michalvavrik** [comment]
I checked every failing test, I think it was all consequence of the broadcasting test. I cannot reproduce it, but when using WSCLient, it seems that some messages just don't arrive (not metrics, messages). I looked how you wrote broadcasting test and rewrote it to use same Vert.x client. Rerun it couple of times locally and it didn't fail. We will see.

File: extensions/websockets-next/deployment/src/test/java/io/quarkus/websockets/next/test/utils/WSClient.java [hunk 7]
```diff
@@ -122,7 +123,7 @@ public Buffer waitForNextMessage() {
     }
 
     public void waitForMessages(int count) {
-        Awaitility.await().until(() -> messages.size() >= count);
+        Awaitility.await().atMost(Duration.ofSeconds(10)).until(() -> messages.size() >= count);
```
Comment (mkouba) [inline extensions/websockets-next/deployment/src/test/java/io/quarkus/websockets/next/test/utils/WSClient.java:126]: Hm, 10 seconds is not that long... esp. for slow machines. I think that we should use longer interval (20?) by default and maybe make it configurable?

**Comment by mkouba** [comment]
> I checked every failing test, I think it was all consequence of the broadcasting test. I cannot reproduce it, but when using WSCLient, it seems that some messages just don't arrive (not metrics, messages). I looked how you wrote broadcasting test and rewrote it to use same Vert.x client. Rerun it couple of times locally and it didn't fail. We will see.

The CI is green so I suppose that your fix worked :+1:

**Review by mkouba — APPROVED** [review]

File: extensions/websockets-next/deployment/src/test/java/io/quarkus/websockets/next/test/utils/WSClient.java (hunk 7 above)
Comment (michalvavrik) [reply extensions/websockets-next/deployment/src/test/java/io/quarkus/websockets/next/test/utils/WSClient.java:126]: I think I have changed this one when I was thinking about 5 second timeout Bruno suggested for waiting for traces (it was in memory handler). I'd not do that if I realized it was method inside WSClient and not the code I was adding.

TL;DR; I can see in `org.awaitility.constraint.AtMostWaitConstraint` that default value when it is reset is 10 seconds so this change is unnecessary. I'll drop it, it's good opportunity to re-run CI again and see if tests are not flaky.

**Comment by brunobat** [comment]
@michalvavrik, I will need more time to review the PR, however, one thing seems clear to me. We need an IT test. 
I think we don't need a new project, just try to add a very simple test on `quarkus-integration-test-opentelemetry`.
This way we will make sure wiring is ok and native mode as well.

**Comment by mkouba** [comment]
If I understand it correctly the only thing we're waiting for is an integration test, right? If so, I'd like to merge this PR and address this last thing in a follow-up.

WDYT? @brunobat @michalvavrik

**Comment by michalvavrik** [comment]
> If I understand it correctly the only thing we're waiting for is an integration test, right? If so, I'd like to merge this PR and address this last thing in a follow-up.

I am fighting with time, but I hope (once again) I could have time to add IT today (I also thought that yesterday :-D). Anyway, I think we are waiting for @brunobat review. Last time we talked Bruno was testing this PR locally and told me he needs more time.

**Comment by michalvavrik** [comment]
> address this last thing in a follow-up.

yeah, that part alone can be done easily separately

**Review by brunobat — COMMENTED** [review]
In general I think the tests are very thorough, however thay are also a bit opaque because of the methods inside `MetricsAsserter`.  Future debugging, might force us to refactor that.

To be fair, all this instrumentation should have been done in the `quarkus-micrometer` project under `io.quarkus.micrometer.runtime.binder`.
Same for OpenTelemetry.
We shouldn't be pulverising the extensions with micrometer and OTel libs because it will make maintenance a pain. 

In any case, having the functionality is more important. Do you think this could be moved in the future?

What do you think @michalvavrik?
... And sorry for not thinking of this before.

**Review by brunobat — APPROVED** [review]
Thanks very much for this @michalvavrik !

**Comment by brunobat** [comment]
> > To be fair, all this instrumentation should have been done in the `quarkus-micrometer` project under `io.quarkus.micrometer.runtime.binder`. Same for OpenTelemetry. We shouldn't be pulverising the extensions with micrometer and OTel libs because it will make maintenance a pain.
> > Do you think this could be moved in the future?
> 
> If I unseal `SendingInterceptor`, `ErrorCountingInterceptor` and `ConnectionInterceptor`, then moving it should be as easy as moving this code to the `quarkus-micrometer` and adding conditional WS-Next dependency to `quarkus-micrometer` or introducing runtime SPI. Same should be true for the OTel integration because I believe there is no tight coupling. I am happy to do that, but maybe it would be easier if we got this in and follow up with moving the code and IT?

Deal!

**Comment by mkouba** [comment]
> and adding conditional WS-Next dependency to `quarkus-micrometer` 

:-1: 

> or introducing runtime SPI.

:+1:
